File: TP3/Node.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <string>
#include <sstream>
#include <iterator>
using namespace std;

#ifndef NODE
#define NODE

class Node
{
public:
    unordered_map<int, int> state;
    int site;
    int type;
    bool is_valid;
    string solution;
    int total_cost;

    Node(int site_count)
    {
        for (int i = 0; i < site_count; i++)
        {
            this->state.insert(make_pair(i, -1));
        }
    }

    Node(unordered_map<int, int> state)
    {
        this->state = state;
        this->solution = "";
        this->site = 0;
        this->type = 0;
        this->is_valid = true;
        this->total_cost = UINT64_MAX;
    }

    Node(unordered_map<int, int> state, int site, int type)
    {
        this->site = site;
        this->type = type;
        this->state = state;
        this->solution = "";
        this->state[site] = type;
    }

    void compute_cost(const vector<vector<int>>& cost_matrix, const unordered_map<int, vector<int>>& graph)
    {
        this->total_cost = 0;
        int index = 0;
        for (auto& it : graph)
        {
            int vertex = it.first;
            vector<int> connected_vertices = it.second;
            if (this->state[it.first] == -1)
            {
                continue;
            }

            for (auto connected_vertex : connected_vertices)
            {
                if (connected_vertex <= index || this->state[connected_vertex] == -1)
                {
                    continue;
                }

                this->total_cost += cost_matrix[this->state[vertex]][this->state[connected_vertex]];
                index++;
            }
        }
    }

    void check_validity(const vector<int>& type_amounts, const int type_count)
    {
        vector<int> site_type_counts = vector<int>(type_count, 0);

        for (size_t i = 0; i < this->state.size(); i++)
        {
            if (this->state[i] != -1)
            {
                site_type_counts[this->state[i]]++;
            }
        }

        for (size_t i = 0; i < site_type_counts.size(); i++)
        {
            if (site_type_counts[i] > type_amounts[i])
            {
                this->is_valid = false;
                return;
            }
        }

        this->is_valid = true;
    }

    string get_solution()
    {
        if (this->solution != "")
        {
            return this->solution;
        }

        string solution;
        for (auto& it : this->state)
        {
            solution += to_string(it.second);
        };

        this->solution = solution;
        return solution;
    }

    bool operator<(const Node& rhs)
    {
        return this->total_cost < rhs.total_cost;
    }

    vector<Node*> create_neighbours(int vertex, vector<int>& edges)
    {
        vector<Node*> nodes;
        for (auto edge : edges)
        {
            if (edge < vertex)
            {
                continue;
            }

            Node* node = new Node(this->state);
            int tmp = node->state[vertex];
            node->state[vertex] = node->state[edge];
            node->state[edge] = tmp;

            if (node->get_solution() != this->get_solution())
            {
                nodes.push_back(node);
            }
        }

        return nodes;
    }
};
#endif

File: TP3/Node.py
from pprint import pprint
import copy
class Node:
    def __init__(self, state: dict, site: int, type: int, is_root=False):
        self.site = site
        self.type = type
        self.state = state
        self.is_root = is_root
        if not self.is_root: 
            self.state[site] = type

    def compute_cost(self, cost_matrix, graph):
        self.total_cost = 0
        index = 0
        for vertex, edges in graph.items():
            if self.state[vertex] is None:
                continue
            
            for edge in edges:
                if edge <= index or self.state[edge] is None:
                    continue
                self.total_cost += cost_matrix[self.state[vertex]][self.state[edge]]
            index += 1

        # for site, type in self.state.items():
        #     if type is None:
        #         edges = graph[site]
        #         for edge in edges:
        #             if self.state[edge] is not None:
        #                 cost = min(cost_matrix[self.state[edge]])
        #                 print(f"min cost for site {site} with edge {edge} is {cost}")
        #                 self.total_cost += cost

    def check_validity(self, type_amounts, type_count):
        site_type_counts = [0 for _ in range(type_count)]
        for _, type in self.state.items():
            if type is not None:
                site_type_counts[type] += 1

        for i in range(len(site_type_counts)):
            if site_type_counts[i] > type_amounts[i]:
                self.valid = False
                return

        self.valid = True

    def create_neighbour(self, k: int, edges: list):
        nodes = []
        for edge in edges:
            if edge < k:
                continue

            node = Node(copy.deepcopy(self.state), 0, 0, is_root=True)
            tmp = node.state[k]
            node.state[k] = node.state[edge]
            node.state[edge] = tmp
            attempting_solution = node.get_solution()
            solution = self.get_solution()

            if attempting_solution != solution:
                nodes.append(node)

        return nodes
        
    def get_solution(self):
        return " ".join([str(type) for _, type in self.state.items()])
        
    def __lt__(self, other_node):
        return self.total_cost < other_node.total_cost
